irispy/spectrograph.py

```diff
@@ -208,6 +209,363 @@ def spectral_windows(self):
+        extra_coords_keys = [coord[0] for coord in extra_coords]
+        if not all([key in extra_coords_keys for key in required_extra_coords_keys]):
+            raise ValueError("The following extra coords must be supplied: {0} vs. {1} from {2}".format(
+                required_extra_coords_keys, extra_coords_keys, extra_coords))
+        # Initialize IRISSpectrogram.
+        super(IRISSpectrogram, self).__init__(
+            data, wcs, uncertainty=uncertainty, mask=mask, meta=meta,
+            unit=unit, extra_coords=extra_coords, copy=copy, missing_axis=missing_axis)
+
+    def __getitem__(self, item):
+        result = super(IRISSpectrogram, self).__getitem__(item)
+        return IRISSpectrogram(
+            result.data, result.wcs, result.uncertainty, result.unit, result.meta,
+            _extra_coords_to_input_format(result._extra_coords, result.missing_axis),
+            mask=result.mask, missing_axis=result.missing_axis)
+
+    def __repr__(self):
+        if self.missing_axis[::-1][self._extra_coords["time"]["axis"]]:
+            instance_start = instance_end = self._extra_coords["time"]["value"]
+        else:
+            instance_start = self._extra_coords["time"]["value"][0],
+            instance_end = self._extra_coords["time"]["value"][-1]
+        return """IRISSpectrogram
+---------------------
+{obs_repr}
+
+Spectrogram period: {inst_start} -- {inst_end}
+Data shape: {shape}
+Axis Types: {axis_types}
+""".format(obs_repr=_produce_obs_repr_string(self.meta),
+           inst_start=instance_start, inst_end=instance_end,
+           shape=self.dimensions, axis_types=self.dimensions.axis_types[::])
+
+    def to(self, new_unit_type):
+        """
+        Converts data, unit and uncertainty attributes to new unit type.
+
+        The presence or absence of the exposure time correction is
+        preserved in the conversions.
+
+        Parameters
+        ----------
+        new_unit_type: `str`
+           Unit type to convert data to.  Three values are accepted:
+           "DN": Relevant IRIS data number based on detector type.
+           "counts": photon counts
+           "radiance": Perorms radiometric calibration conversion.
+
+        Returns
+        -------
+        result: `IRISSpectrogram`
+            New IRISSpectrogram in new units.
+
+        """
+        detector_type = iris_tools.get_detector_type(self.meta)
+        if new_unit_type == "radiance" or self.unit.is_equivalent(iris_tools.RADIANCE_UNIT):
+            # Get spectral dispersion per pixel.
+            spectral_wcs_index = np.where(np.array(self.wcs.wcs.ctype) == "WAVE")[0][0]
+            spectral_dispersion_per_pixel = self.wcs.wcs.cdelt[spectral_wcs_index] * \
+                                            self.wcs.wcs.cunit[spectral_wcs_index]
+            # Get solid angle from slit width for a pixel.
+            lat_wcs_index = np.where(np.array(self.wcs.wcs.ctype) == "HPLT-TAN")[0][0]
```

Comment:
would probably be good to only check for `HPLT` here not the projection as well.